### DISCUSSION TITLE
Local development workflow

### DISCUSSION THREAD

**Discussion by waza-ari** [copier-org/copier#1626] [Q&A]
First of all, thanks for creating this awesome project! I'm currently using it to offer (yet another) FastAPI boilerplate template. In general that works perfectly fine, my question is around local development.

A few important files are templated, e.g. `pyproject.toml`, `main.py` and so on, so they're instead called `pyproject.toml.jinja` and `main.py.jinja` respectively. To run, develop and extend the template locally, I need to have those files available though. So far, I'm just having a copy of those files and have the actual file names in `.gitignore` to prevent accidental commits.

This however would result in a problem after using the template, as the `gitignore` file would still contain the files which have been created from the template, effectively meaning that the user would need to manually change the `.gitignore` file. I assume this would result in problems with updates.

What would be the recommended workflow to support such scenarios?

**Comment by pawamoy** [comment]
Could you post snippets of your file layout? This would help tremendously understanding how you're using Copier. It's very hard to do that from text alone :slightly_smiling_face:

**Reply by waza-ari** [reply]
Hi, thanks for your answer, and my apologies for not coming back earlier. I somehow missed that :( The project (very early stage) is on Github: https://github.com/waza-ari/fastapi-template

In the root directory, let's look at [pyproject.toml.jinja](https://github.com/waza-ari/fastapi-template/blob/main/pyproject.toml.jinja) as an example. It needs to be templated in order to use the project. When working with the template itself (like continuing the development) locally, I also need that file available though.

So typically, in my local development version, there are two files:

- pyproject.toml.jinja   <-- the one that is also committed into Giit
- pyproject.toml           <-- derived from the first, but I need it for local development

Now Git obviously detects `pyproject.toml` as a new file and wants to add it to the repo. Adding it to gitignore is not an option as it would make the file ignored **after** templating for the user of the project.

It seems like a pretty common scenario to me, so I'm pretty sure I'm missing something super obvious

**Comment by yajo** [comment]
It seems to me you're mixing the template and the subproject.

In the template, you only need `pyproject.toml.jinja`, as the template's sole purpose in life is to generate subprojects. Templates are handled by copier (not by pip); therefore they don't _need_ a `pyproject.toml`. The template has its own git repo.

Then you use this to generate your subproject:

    copier copy https://github.com/waza-ari/fastapi-template ~/subproject

Then you use that template to render it into a subproject. The subproject will indeed need a `pyproject.toml` file, but it won't need a `pyproject.toml.jinja` file because that's living inside the template. The subproject is a real python package, and it only needs python code, not Copier code.
